Documentation and Guides (#164)
fiberplane/fiberplane: www/src/content/docs/components/Otel-middleware.mdx

---

> title: Otel middleware

Author: high-level question for discussion:

What should we call this component? 

Technically it's not middleware in the Hono sense, and we know from experience that "Otel" is not familiar to many devs. Maybe we can keep it to speak to the otel-initiated, but. Worth revisiting the name in the sidebar

My point kinda being: "Otel middleware" feels a little bit like a misnomer.

A few alts off the top of my head:

- Client instrumentation
- Client library
- Hono client library
- Hono otel library 

None of them is great. But. A start at the very least

Reviewer: Hono instrumentation library ?

Reviewer: I like if it is specific and contains somewhere Otel or instrumentation.

Reviewer 2: I like the "Client library". It tells the user in the most broad sense of what this is (a client library to the studio), without expecting them to understand what OTEL and "instrumentation" are at the top of the hour. We do of course explain what those are in the body of the text.

description: description of otel middleware
---

> The Hono [Otel middleware](https://github.com/fiberplane/fpx/tree/main/packages/client-library-otel) leverages [OpenTelemetry](https://opentelemetry.io/) to instrument applications and send traces to FPX Studio. It requires the use of [Hono](https://hono.dev/) web application framework.

Author: ```suggestion
The Hono [Otel middleware](https://github.com/fiberplane/fpx/tree/main/packages/client-library-otel) leverages [OpenTelemetry](https://opentelemetry.io/) to instrument applications and send traces to FPX Studio. It requires use of the [Hono](https://hono.dev/) web application framework.
```

## Installation 
To install the Otel middleware for your Hono application, run the following command in your project directory:
```
npm i @fiberplane/hono-otel@beta 
``` 

> ## Usage within the Hono application

Author: ```suggestion
## Usage within a Hono application
```

Below is an example of how to use the Otel middleware within a Hono application:
```typescript
import { Hono } from "hono";
import { instrument } from "@fiberplane/hono-otel";

const app = new Hono();

app.get("/", (c) => c.text("Hello, Hono!"));

export default instrument(app);
```

### Advanced usage: Custom spans

> The measure function from the Hono Otel middleware allows you to create custom spans for any function within your application, enabling more granular tracing.

Author: Question for discussion: Can/should we link to something like "what is a span" or "what is a trace"?

Reviewer: what about this link form the Otel docs: https://opentelemetry.io/docs/concepts/signals/traces/

Reviewer 2: I think we can link for now but ideally we'd explain those on our site, without having to link off elsewhere for the core explanations (linking for source and in-depth explorations ofc).

```typescript
import { instrument, measure } from "@fiberplane/hono-otel";

const app = new Hono();

// Create a loop function that will get recorded as a span inside the trace for a incoming given request
const loop = measure("loop", (n: number) => {
  for (let i = 0; i < n; i++) {
    console.log(`Loop iteration: ${i}`);
  }
});

app.get("/", (c) => {
  loop(100);
  return c.text("Hello, Hono!");
});

export default instrument(app);
```

## Configuration

> By default, the Otel Hono middleware does not send traces. To define an endpoint, you need to set the `FPX_ENDPOINT` environment variable.

Author: ```suggestion
By default, the Hono Otel middleware does not send traces. To define an endpoint, you need to set the `FPX_ENDPOINT` environment variable.
```

When starting Fiberplane Studio via the command line with `npx @fiberplane/studio`, the `FPX_ENDPOINT` is automatically set to the Fiberplane Studio's default port 8788: `http://localhost:8788/v1/traces`. Additionally, the Fiberplane Studio CLI setup enables Node.js compatibility mode for Cloudflare Workers.

> If you are not using Fiberplane Studio, ensure that Node.js compatibility mode is enabled in the `wrangler.toml` file:

Author: Worth noting: `wrangler.toml` is specific to Cloudflare setups

Reviewer: yes true, I will add a line saying it is only needed for cloudflare workers

```
compatibility_flags = [ "nodejs_compat" ]
``` 
### Customizing the `instrument` function behavior

> The middleware allows you to override the default behavior of the `instrument` function. By default, the function creates traces for all fetched requests in the application and sends logging data to a locally running FPX Studio. You can disable these features by setting the corresponding monitoring variables to `false`. 

Author: ```suggestion
The middleware allows you to override the default behavior of the `instrument` function. By default, the function creates traces for all fetch requests in the application, and sends logging data to a locally running Fiberplane Studio. You can disable these features by setting the corresponding monitoring variables to `false`. 
```

```typescript
import { Hono } from "hono";
import { instrument } from "@fiberplane/hono-otel";

const app = new Hono();

app.get("/", (c) => c.text("Hello, Hono!"));

export default instrument(app, {
  monitor: {
    // Don't create traces for fetch requests
    fetch: false,

>     // Don't proxy `console.*` functions to send logging data to a local FPX server

Author: ```suggestion
    // Don't proxy `console.*` functions to send log events
```

    logging: false,
  },
});
```

> Find the code on [Github](https://github.com/fiberplane/fpx/tree/docs/www)

Author: is this "Find the code" intended to link to this page specifically?

if we have this "Find the code" link on each page, we could implement a custom footer to add it to each page

Reviewer: no sorry the link is wrong. I wanted to link the Otel library but i think this get messed up. I think for studio and the otel client library it is good to include a link to the code on github. i don't think we need it for the other docs elements
